optaplanner-core/src/main/java/org/optaplanner/core/api/score/holder/AbstractScoreHolder.java

```diff
@@ -235,11 +235,21 @@ private BigDecimalConstraintMatchTotal findBigDecimalConstraintMatchTotal(RuleCo
         return matchTotal;
     }
 
-    private AgendaItem prepareAgendaItemForUnMatchListener(RuleContext kcontext) {
+    private AgendaItem prepareAgendaItemForUnMatchListener(RuleContext kcontext, int scoreLevel) {
         AgendaItem agendaItem = (AgendaItem) kcontext.getMatch();
-        if (agendaItem.getActivationUnMatchListener() != null) {
-            // Both parameters null because they are not used by the ActivationUnMatchListener created anyway
-            agendaItem.getActivationUnMatchListener().unMatch(null, null);
+        ConstraintMatchTotal matchTotal = null;
+        if (constraintMatchEnabled) {
+            Rule rule = kcontext.getRule();
+            matchTotal = constraintMatchTotalMap.get(Arrays.asList(rule.getPackageName(), rule.getName(), scoreLevel));
+            if (agendaItem.getActivationUnMatchListener() != null && matchTotal != null) {
+                // Both parameters null because they are not used by the ActivationUnMatchListener created anyway
+                agendaItem.getActivationUnMatchListener().unMatch(null, null);
+            }
+        } else {
+            if (agendaItem.getActivationUnMatchListener() != null) {
+                // Both parameters null because they are not used by the ActivationUnMatchListener created anyway
+                agendaItem.getActivationUnMatchListener().unMatch(null, null);
+            }
```

Comment:
I don't see how this fixes the bug, especially if constraintMatch is not enabled?

HardSoftScoreHolder.addHardConstraintMatch(...) calls registerIntConstraintMatch() which calls:
  agendaItem.setActivationUnMatchListener(constraintUndoListener);
It's important to note the "set" in that drools method "setActivationUnMatchListener". It means any existing ActivationUnMatchListener is removed.

So if a user calls addHardConstraintMatch(...) and addSoftconstraintMatch(), they will still uising the same agendaItem, so last-one-wins: only the last one's ActivationUnMatchListener actually works.